### ISSUE TITLE
[BUG] XSLT Processing Instruction Causes HTTP ERROR 500 javax.servlet.ServletException

### ISSUE THREAD

**Issue by sgmlguru** [eXist-db/exist#4691]
**Describe the bug**
An XSL processing instruction pointing to an XSLT before an XML root element, for example, `<?xml-stylesheet href="../../xsltforms/xsltforms.xsl" type="text/xsl"?>` causes HTTP ERROR 500 javax.servlet.ServletException:

```
URI:	/exist/apps/xsltforms-demo/data/spreadsheet.xml
STATUS:	500
MESSAGE:	javax.servlet.ServletException: javax.servlet.ServletException: An error occurred while processing request to /exist/apps/xsltforms-demo/data/spreadsheet.xml: An error occurred: null
SERVLET:	XQueryURLRewrite
CAUSED BY:	javax.servlet.ServletException: javax.servlet.ServletException: An error occurred while processing request to /exist/apps/xsltforms-demo/data/spreadsheet.xml: An error occurred: null
CAUSED BY:	javax.servlet.ServletException: An error occurred while processing request to /exist/apps/xsltforms-demo/data/spreadsheet.xml: An error occurred: null
CAUSED BY:	javax.servlet.ServletException: An error occurred: null
CAUSED BY:	java.util.EmptyStackException

javax.servlet.ServletException: javax.servlet.ServletException: An error occurred while processing request to /exist/apps/xsltforms-demo/data/spreadsheet.xml: An error occurred: null
	at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:162)
	at org.eclipse.jetty.server.handler.gzip.GzipHandler.handle(GzipHandler.java:772)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.Server.handle(Server.java:516)
	at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:487)
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:732)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:479)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:277)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
	at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:338)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:315)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:173)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:131)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:409)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:883)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1034)
	at java.lang.Thread.run(Thread.java:750)
Caused by: javax.servlet.ServletException: An error occurred while processing request to /exist/apps/xsltforms-demo/data/spreadsheet.xml: An error occurred: null
	at org.exist.http.urlrewrite.XQueryURLRewrite.service(XQueryURLRewrite.java:366)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder$NotAsync.service(ServletHolder.java:1450)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:799)
	at org.eclipse.jetty.servlet.ServletHandler$ChainEnd.doFilter(ServletHandler.java:1656)
	at org.eclipse.jetty.websocket.server.WebSocketUpgradeFilter.doFilter(WebSocketUpgradeFilter.java:292)
	at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:193)
	at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1626)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:552)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:571)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:235)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1624)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1440)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:505)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1594)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1355)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:234)
	at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:146)
	... 18 more
Caused by: javax.servlet.ServletException: An error occurred: null
	at org.exist.http.servlets.EXistServlet.doGet(EXistServlet.java:363)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:687)
	at org.exist.http.servlets.EXistServlet.service(EXistServlet.java:588)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder$NotAsync.service(ServletHolder.java:1459)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:799)
	at org.eclipse.jetty.servlet.ServletHandler$ChainEnd.doFilter(ServletHandler.java:1656)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:552)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:618)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:235)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1624)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1440)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:505)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1594)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1355)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.Dispatcher.forward(Dispatcher.java:167)
	at org.eclipse.jetty.server.Dispatcher.forward(Dispatcher.java:81)
	at org.exist.http.urlrewrite.Forward.doRewrite(Forward.java:50)
	at org.exist.http.urlrewrite.XQueryURLRewrite.doRewrite(XQueryURLRewrite.java:523)
	at org.exist.http.urlrewrite.XQueryURLRewrite.service(XQueryURLRewrite.java:340)
	... 41 more
Caused by: java.util.EmptyStackException
	at java.util.Stack.peek(Stack.java:102)
	at net.sf.saxon.style.UseWhenFilter.processBaseUri(UseWhenFilter.java:433)
	at net.sf.saxon.style.UseWhenFilter.startElement(UseWhenFilter.java:140)
	at net.sf.saxon.event.StartTagBuffer.startContent(StartTagBuffer.java:246)
	at net.sf.saxon.event.ProxyReceiver.startContent(ProxyReceiver.java:177)
	at net.sf.saxon.event.ProxyReceiver.startContent(ProxyReceiver.java:177)
	at net.sf.saxon.event.ReceivingContentHandler.startElement(ReceivingContentHandler.java:446)
	at org.exist.util.serializer.ReceiverToSAX.startElement(ReceiverToSAX.java:105)
	at org.exist.storage.serializers.NativeSerializer.serializeToReceiver(NativeSerializer.java:273)
	at org.exist.storage.serializers.NativeSerializer.serializeToReceiver(NativeSerializer.java:132)
	at org.exist.storage.serializers.Serializer.setStylesheet(Serializer.java:766)
	at org.exist.storage.serializers.Serializer.toSAX(Serializer.java:853)
	at org.exist.http.RESTServer.writeResourceAs(RESTServer.java:1816)
	at org.exist.http.RESTServer.doGet(RESTServer.java:417)
	at org.exist.http.servlets.EXistServlet.doGet(EXistServlet.java:324)
	... 66 more
```

This particular example happened when trying to open the XForms example spreadsheet.xml in the XSLTForms Demo application, v 1.3.

**Expected behavior**
The XML should open as an XForms document by the XSLTForms application.

**To Reproduce**

1. Install XSLTForms v 1.3 and XSLTForms Demo v 1.3 from the Package Manager.
2. Open the XSLTForms Demo application from the Dashboard.
3. Open the [Spreadsheet example](http://localhost:8080/exist/apps/xsltforms-demo/data/spreadsheet.xml). This will result in the above error.
4. Any other XForm XML with a PI to reference XSLTForms will fail in a similar way.

**Screenshots**
N/A

**Context (please always complete the following information):**
 - OS: Ubuntu Linux 22.10
 - eXist-db version: 6.1.0 (released 15 Jan 23)
 - Java Version OpenJDK 1.8.0_352-8u352-ga-1~22.10-b08

**Additional context**
- How is eXist-db installed? JAR
- Any custom changes in e.g. `conf.xml`? No changes
- Same issue with 5.5.1, released on 15 Jan

**Comment by dizzzz** [comment]
I guess this is a Saxon issue (handled too rough by eXist) ?

**Comment by adamretter** [comment]
Thanks @sgmlguru with your instructions I was able to easily reproduce it.

What I have discovered is that the RESTServer processes the XSLT declared by the xsl-pi twice:

1. `RESTServer#writeResourceAs` -> `Serializer#prepareStylesheets` -> `Serializer#setStylesheet`
2. `RESTServer#writeResourceAs` ->  `Serializer#toSAX` -> `Serializer#setStylesheet`

The interesting thing is that the first time the XSLT is processed that succeeds just fine. The second time however causes the error (presumably due to some state not being reset before the second call).
It seems to me however that the RESTServer should not cause the XSLT to be processed twice, as it need be done only once - so this is quite likely the bug.

I will try and find some more time to dig in further soon - unfortunately this part of the eXist-db code is not-documented and not-obvious...
